internal/login/login.go

```diff
@@ -1,18 +1,185 @@
+	if err != nil {
+		return "", fmt.Errorf("%s: %w", decryptionErrorMsg, err)
+	}
+
+	block, err := aes.NewCipher(secret)
+	if err != nil {
+		return "", fmt.Errorf("%s: %w", decryptionErrorMsg, err)
+	}
+
+	aesgcm, err := cipher.NewGCM(block)
+	if err != nil {
+		return "", fmt.Errorf("%s: %w", decryptionErrorMsg, err)
+	}
+
+	decryptedAccessToken, err := aesgcm.Open(nil, nonce, decodedAccessToken, nil)
+	if err != nil {
+		return "", fmt.Errorf("%s: %w", decryptionErrorMsg, err)
+	}
+
+	return string(decryptedAccessToken), nil
+}
+
+func pollForAccessToken(url string) (AccessTokenResponse, error) {
+	var accessTokenResponse AccessTokenResponse
+
+	// We fully control the url here, so it's safe to perform the request and ignore the G107 gosec rule.
+	resp, err := http.Get(url) //nolint:gosec
+	if err != nil {
+		return accessTokenResponse, fmt.Errorf("cannot fetch access token: %w", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode == http.StatusNotFound {
+		retryAfterSeconds, err := strconv.Atoi(resp.Header.Get("Retry-After"))
+		if err != nil {
+			retryAfterSeconds = defaultRetryAfterSeconds
+		}
+		time.Sleep(time.Duration(retryAfterSeconds) * time.Second)
+		return pollForAccessToken(url)
+	}
+
+	if resp.StatusCode == http.StatusOK {
+		body, err := io.ReadAll(resp.Body)
+
+		if err != nil {
+			return accessTokenResponse, fmt.Errorf("cannot read access token response body: %w", err)
+		}
+
+		if err := json.Unmarshal(body, &accessTokenResponse); err != nil {
+			return accessTokenResponse, fmt.Errorf("cannot unmarshal access token response: %w", err)
+		}
+
+		return accessTokenResponse, nil
+	}
+
+	return accessTokenResponse, fmt.Errorf("HTTP %s: cannot retrieve access token", resp.Status)
+}
+
+func Run(stdin *os.File, params RunParams) error {
+	if params.Token != "" {
+		return utils.SaveAccessToken(params.Token, params.Fsys)
+	}
+
+	if params.OpenBrowser {
+		fmt.Print("Hello from ", utils.Aqua("Supabase"), "! Press ", utils.Aqua("Enter"), " to open browser and login automatically.\n")
+		fmt.Scanln()
+	}
+
+	sessionId := uuid.New().String()
+	tokenName := params.Name
+	curve := ecdh.P256()
+	privateKey, err := curve.GenerateKey(rand.Reader)
+	if err != nil {
+		return fmt.Errorf("cannot generate encryption key: %w", err)
+	}
+	publicKey := privateKey.PublicKey()
+	encodedPublicKey := hex.EncodeToString(publicKey.Bytes())
+
+	createLoginSessionPath := "/cli/login"
+	createLoginSessionQuery := "?session_id=" + sessionId + "&token_name=" + tokenName + "&public_key=" + encodedPublicKey
+	createLoginSessionUrl := utils.GetSupabaseDashboardURL() + createLoginSessionPath + createLoginSessionQuery
+
+	if params.OpenBrowser {
+		fmt.Printf("Here is your login link in case browser did not open: %s\n\n", utils.Bold(createLoginSessionUrl))
+	} else {
+		fmt.Printf("Here is your login link: %s\n\n", utils.Bold(createLoginSessionUrl))
+	}
+
+	spinner := spinner.New(spinner.CharSets[11], 100*time.Millisecond)
```

Comment:
We have been using bubbletea/spinner for existing commands like [supabase stop](https://github.com/supabase/cli/blob/main/internal/stop/stop.go#L26). Could you try if that works?

Reply:
Updated. Although it's now quite obvious why I went with other library for such a simple UI thing :p

https://github.com/supabase/cli/pull/1630/commits/127b9e2723f7294a69e7a4db1892203639155ff3#diff-63f7df4adcd885e26af5017d4653902be40956c5d99878eb9795b1c406b16d77

Reply:
Tbh we can also remove it entirely, or redo `RunProgram` to return the value instead of going through channels